frontend/src/Utils/api-discovery.ts

```diff
@@ -0,0 +1,212 @@
+import { commonFetchJSON, K8sGroupVersionKind, K8sModelCommon } from '@openshift/dynamic-plugin-sdk-utils';
```

Comment:
Diff of this file with what we have in `lib-utils` API discovery today:

```diff
1,3c1,8
< import type { AnyObject } from '@openshift/dynamic-plugin-sdk';
< import { consoleLogger } from '@openshift/dynamic-plugin-sdk';
< import * as _ from 'lodash-es';
---
> import { commonFetchJSON, K8sGroupVersionKind, K8sModelCommon } from '@openshift/dynamic-plugin-sdk-utils';
> import { ThunkDispatch } from 'redux-thunk';
> import { AnyAction } from 'redux';
> import mergeWith from 'lodash/mergeWith';
> import merge from 'lodash/merge';
> import keyBy from 'lodash/keyBy';
> import chunk from 'lodash/chunk';
> import startCase from 'lodash/startCase';
5,7c10,13
< import type { Dispatch } from 'redux';
< import { kindToAbbr } from '../../k8s/k8s-resource';
< import type {
---
> import { receivedResources, setResourcesInFlight, setBatchesInFlight } from '../store/actions';
> import {
>   AnyObject,
>   APIResourceData,
9a16
>   GroupVersion,
11,16c18,20
< } from '../../types/api-discovery';
< import type { K8sModelCommon } from '../../types/k8s';
< import type { DispatchWithThunk } from '../../types/redux';
< import { commonFetchJSON } from '../../utils/common-fetch';
< import { setResourcesInFlight, setBatchesInFlight, receivedResources } from '../redux/actions/k8s';
< import { cacheResources, getCachedResources } from './discovery-cache';
---
>   K8sResourceKindReference,
>   SDKStoreState,
> } from './api-discovery.types';
19a24,25
> const SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY = 'sdk/api-discovery-resources';
> const abbrBlacklist = ['ASS'];
20a27,31
> export const kindToAbbr = (kind: string) => {
>   const abbrKind = (kind.replace(/[^A-Z]/g, '') || kind.toUpperCase()).slice(0, 4);
>   return abbrBlacklist.includes(abbrKind) ? abbrKind.slice(0, -1) : abbrKind;
> };
> 
23c34
<   const pluralized = plural(_.startCase(kind)).replace(/\s+/g, '');
---
>   const pluralized = plural(startCase(kind)).replace(/\s+/g, '');
30a42,103
> const getCachedResources = () => {
>   const resourcesJSON = localStorage.getItem(SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY);
>   if (!resourcesJSON) {
>     // eslint-disable-next-line no-console
>     console.error(`No API resources found in localStorage for key ${SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY}`);
>     return null;
>   }
> 
>   // Clear cached resources after load as a safeguard. If there's any errors
>   // with the content that prevents the console from working, the bad data
>   // will not be loaded when the user refreshes the console. The cache will
>   // be refreshed when discovery completes.
>   localStorage.removeItem(SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY);
> 
>   const resources = JSON.parse(resourcesJSON);
>   // eslint-disable-next-line no-console
>   console.info('Loaded cached API resources from localStorage');
>   return resources;
> };
> 
> const getLocalResources = () => {
>   try {
>     return JSON.parse(localStorage.getItem(SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY) || '{}');
>   } catch (e) {
>     // eslint-disable-next-line no-console
>     console.error('Cannot load cached API resources', e);
>     throw e;
>   }
> };
> 
> export const getReference = ({ group, version, kind }: K8sGroupVersionKind): K8sResourceKindReference => [group || 'core', version, kind].join('~');
> 
> export const getReferenceForModel = (model: K8sModelCommon): K8sResourceKindReference =>
>   getReference({ group: model.apiGroup, version: model.apiVersion, kind: model.kind });
> 
> const mergeByKey = (prev: K8sModelCommon[], next: K8sModelCommon[]) =>
>   Object.values(merge(keyBy(prev, getReferenceForModel), keyBy(next, getReferenceForModel)));
> 
> const setLocalResources = (resources: DiscoveryResources[]) => {
>   try {
>     localStorage.setItem(SDK_API_DISCOVERY_RESOURCES_LOCAL_STORAGE_KEY, JSON.stringify(resources));
>   } catch (e) {
>     // eslint-disable-next-line no-console
>     console.error('Error caching API resources in localStorage', e);
>     throw e;
>   }
> };
> export const cacheResources = (resources: DiscoveryResources[]) => {
>   const allResources = [...[getLocalResources()], ...resources].reduce(
>     (acc, curr) =>
>       mergeWith(acc, curr, (first: K8sModelCommon[], second: K8sModelCommon[]) => {
>         if (Array.isArray(first) && first[0]?.constructor?.name === 'Object') {
>           return mergeByKey(first, second);
>         }
> 
>         return undefined;
>       }),
>     {},
>   );
>   setLocalResources(allResources);
> };
> 
59,72c132
< type APIResourceData = {
<   groups: {
<     name: string;
<     versions: {
<       groupVersion: unknown;
<     }[];
<     preferredVersion: { version: unknown };
<   }[];
< };
< 
< const batchResourcesRequest = (
<   batch: string[],
<   groupVersionMap: AnyObject,
< ): Promise<DiscoveryResources>[] => {
---
> const batchResourcesRequest = (batch: string[]): Promise<DiscoveryResources>[] => {
87,107c147,151
<     const models = _.flatten(defineModels(resourceList));
<     const coreResources = new Set([
<       'roles',
<       'rolebindings',
<       'clusterroles',
<       'clusterrolebindings',
<       'thirdpartyresources',
<       'nodes',
<       'secrets',
<     ]);
<     allResources.forEach((r) =>
<       coreResources.has(r.split('/')[0]) ? adminResources.push(r) : safeResources.push(r),
<     );
<     const configResources = _.filter(
<       models,
<       (m) => m.apiGroup === 'config.openshift.io' && m.kind !== 'ClusterOperator',
<     );
<     const clusterOperatorConfigResources = _.filter(
<       models,
<       (m) => m.apiGroup === 'operator.openshift.io',
<     );
---
>     const models = defineModels(resourceList).flat();
>     const coreResources = new Set(['roles', 'rolebindings', 'clusterroles', 'clusterrolebindings', 'thirdpartyresources', 'nodes', 'secrets']);
>     allResources.forEach((r) => (coreResources.has(r.split('/')[0]) ? adminResources.push(r) : safeResources.push(r)));
>     const configResources = models.filter((m) => m.apiGroup === 'config.openshift.io' && m.kind !== 'ClusterOperator');
>     const clusterOperatorConfigResources = models.filter((m) => m.apiGroup === 'operator.openshift.io');
117d160
<       groupVersionMap,
122,125c165
< const getResources = async (
<   preferenceList: string[],
<   dispatch: Dispatch,
< ): Promise<DiscoveryResources> => {
---
> const getResources = async (preferenceList: string[]): Promise<[(string | string[])[][], GroupVersion]> => {
127,128c167
<   const groupVersionMap = apiResourceData.groups.reduce(
<     (acc: AnyObject, { name, versions, preferredVersion: { version } }) => {
---
>   const groupVersionMap = apiResourceData.groups.reduce((acc: AnyObject, { name, versions, preferredVersion: { version } }) => {
130c169
<         versions: _.map(versions, 'version'),
---
>       versions: versions.map(({ version: oneVersion }) => oneVersion),
134,143c173,181
<     },
<     {},
<   );
<   const all = ['/api/v1'].concat(
<     _.flatten(
<       apiResourceData.groups.map<string[]>((group) =>
<         group.versions.map<string>((version) => `/apis/${version.groupVersion}`),
<       ),
<     ).sort((api) => (preferenceList.find((item) => api.includes(`/apis/${item}`)) ? -1 : 0)),
<   );
---
>   }, {});
>   const all = [
>     '/api/v1',
>     ...apiResourceData.groups
>       .flatMap<string[]>((group) => group.versions.map<string>((version) => `/apis/${version.groupVersion}`))
>       .sort((api) => (preferenceList.find((item) => api.includes(`/apis/${item}`)) ? -1 : 0)),
>   ];
>   return [chunk(all, API_DISCOVERY_REQUEST_BATCH_SIZE), groupVersionMap as GroupVersion];
> };
145,148c183,187
<   // let batchedData: APIResourceList[] = [];
<   const batches = _.chunk(all, API_DISCOVERY_REQUEST_BATCH_SIZE);
<   const allResources: DiscoveryResources[] = [];
<   // eslint-disable-next-line no-restricted-syntax
---
> const startAPIDiscovery = async (preferenceList: string[], dispatch: ThunkDispatch<SDKStoreState, undefined, AnyAction>) => {
>   try {
>     dispatch(setResourcesInFlight(true));
>     dispatch(setBatchesInFlight(true));
>     const [batches, groupVersionMap] = await getResources(preferenceList);
150,157c189,191
<     // eslint-disable-next-line no-await-in-loop
<     const resources = await Promise.all(batchResourcesRequest(batch, groupVersionMap));
<     resources.forEach((resource) => {
<       allResources.push(resource);
<       dispatch(receivedResources(resource));
<     });
<     // Cache the resources whenever discovery completes to improve console load times.
<     cacheResources(resources);
---
>       const resources = await Promise.all(batchResourcesRequest(batch as string[]));
>       cacheResources(resources as DiscoveryResources[]);
>       resources.map((resource: DiscoveryResources) => dispatch(receivedResources({ ...resource, ...groupVersionMap })));
159d192
<   // Dispatch action to indicate all batches were loaded
161c194,198
<   return allResources.reduce((acc, curr) => _.merge(acc, curr));
---
>     dispatch(setResourcesInFlight(false));
>   } catch (e) {
>     // eslint-disable-next-line no-console
>     console.error('API discovery startAPIDiscovery failed:', e);
>   }
164,183d200
< const updateResources =
<   (preferenceList: string[]) =>
<   async (dispatch: Dispatch): Promise<DiscoveryResources> => {
<     dispatch(setResourcesInFlight(true));
<     dispatch(setBatchesInFlight(true));
< 
<     const resources = await getResources(preferenceList, dispatch);
< 
<     return resources;
<   };
< 
< const startAPIDiscovery = (preferenceList: string[]) => (dispatch: DispatchWithThunk) => {
<   dispatch(updateResources(preferenceList))
<     .then((resources) => {
<       return resources;
<     })
<     // TODO handle failures - retry if error is recoverable
<     .catch((err) => consoleLogger.error('API discovery startAPIDiscovery failed:', err));
< };
< 
190c207,208
<   consoleLogger.info(`API discovery waiting ${API_DISCOVERY_INIT_DELAY} ms before initializing`);
---
>   // eslint-disable-next-line no-console
>   console.info(`API discovery waiting ${API_DISCOVERY_INIT_DELAY} ms before initializing`);
192c210
<     storeInstance.dispatch(startAPIDiscovery(preferenceList));
---
>     startAPIDiscovery(preferenceList, storeInstance.dispatch);
```

Reply:
Ah. I see we combined the cache and the implementation  into one file.

Reply:
Yeah, it's not a carbon copy of what's in lib-utils, but slightly adjusted.